File: backend-projeto-tarefa/src/main/java/service/TarefaService.java
package service;

import java.util.ArrayList;
import java.util.List;

import exception.TarefaException;
import model.dto.TemplateTarefaDTO;
import model.entity.ItemTarefa;
import model.entity.Tarefa;
import model.repository.ItemTarefaRepository;
import model.repository.TarefaRepository;
import model.seletor.TarefaSeletor;

public class TarefaService {

	private TarefaRepository tarefaRepository = new TarefaRepository();
	private ItemTarefaRepository itemRepository = new ItemTarefaRepository();

	public Tarefa inserir(Tarefa novaTarefa) throws TarefaException {
		this.validarCamposObrigatorios(novaTarefa);
		return tarefaRepository.inserir(novaTarefa);
	}

	public Tarefa consultarPorId(int id) {
		return tarefaRepository.consultarPorId(id);
	}

	public ArrayList<Tarefa> consultarTodos() {
		return tarefaRepository.consultarTodos();
	}

	public boolean alterar(Tarefa alterarTarefa) {
		return tarefaRepository.alterar(alterarTarefa);
	}

	public boolean excluir(int id) throws TarefaException {
		if (tarefaRepository.consultarPorId(id) != null && tarefaRepository.consultarPorId(id).getRealizado()) {
			return tarefaRepository.excluir(id);
		} else {
			throw new TarefaException("Tarefa não foi realizada, portanto não pode ser excluídos!");
		}
	}

	public ArrayList<Tarefa> consultarPorFiltro(TarefaSeletor seletor) {
		return tarefaRepository.consultarPorFiltro(seletor);
	}

	public int contarTotalRegistro(TarefaSeletor seletor) {
		return this.tarefaRepository.contarTotalDeRegistro(seletor);
	}

	public int contarPaginas(TarefaSeletor seletor) {
		return this.tarefaRepository.contarPaginas(seletor);
	}

	public List<Tarefa> listaTemplate() {
		return tarefaRepository.listarTemplates();
	}

	public Tarefa criarTarefaAPartirDeTemplate(TemplateTarefaDTO templateDTO) throws TarefaException {

		Tarefa tarefaTemplate = tarefaRepository.consultarPorId(templateDTO.getIdTarefaTemplate());

		if (tarefaTemplate == null || tarefaTemplate.getIsTemplate() == false) {
			throw new TarefaException("Tarefa não é um template");
		}

		// Cria nova tarefa
		Tarefa novaTarefa = new Tarefa();
		novaTarefa.setUsuario(tarefaTemplate.getUsuario());
		novaTarefa.setTipoTarefa(tarefaTemplate.getTipoTarefa());
		novaTarefa.setNomeTarefa(templateDTO.getNomeNovaTarefa());
		novaTarefa = tarefaRepository.inserir(novaTarefa);

		// Cria os itens da nova tarefa
		for (ItemTarefa itemTarefaTemplate : tarefaTemplate.getItensTarefa()) {
			ItemTarefa novoItem = new ItemTarefa();
			novoItem.setDescricao(itemTarefaTemplate.getDescricao());
			novoItem.setTarefa(novaTarefa);

			itemRepository.inserir(novoItem);
		}

		// Chamado para atualizar a lista de itens
		novaTarefa = tarefaRepository.consultarPorId(novaTarefa.getIdTarefa());
		return novaTarefa;
	}

	public Tarefa concluirTarefa() {

		Tarefa tarefa = new Tarefa();
		ArrayList<ItemTarefa> list = itemRepository.consultarItensPedentes(tarefa.getIdTarefa());

		for (ItemTarefa itemTarefa : list) {
			itemRepository.marcarItemComoRealizado(itemTarefa.getIdItem());
		}

		tarefa = tarefaRepository.consultarPorId(tarefa.getIdTarefa());

		return tarefa;

	}

	private void validarCamposObrigatorios(Tarefa t) throws TarefaException {

		String mensagemValidacao = "";
		if (t.getNomeTarefa() == null || t.getNomeTarefa().isEmpty()) {
			mensagemValidacao += " - informe o nome \n";
		}
		if (t.getTipoTarefa() == null || t.getTipoTarefa().isEmpty()) {
			mensagemValidacao += " - informe o tipo da tarefa \n";
		}
		if (!mensagemValidacao.isEmpty()) {

			throw new TarefaException("Preencha o(s) seguinte(s) campo(s) \n " + mensagemValidacao);
		}
	}

}

File: projeto-tarefa-frontend/src/app/shared/service/tarefa.service.ts
import { Injectable } from '@angular/core';
import { Tarefa } from '../model/tarefa';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { TarefaSeletor } from '../model/seletor/tarefaSeletor';
import { TarefaTemplateDTO } from '../model/DTO/TarefaTemplateDTO';
import { ItemTarefa } from '../model/itemTarefa';

@Injectable({
  providedIn: 'root'
})
export class TarefaService {

  private readonly API = 'http://localhost:8080/rest/tarefa';

  constructor(private httpClient: HttpClient) { }


  public inserir(tarefa: Tarefa): Observable<Tarefa> {
    return this.httpClient.post<Tarefa>(this.API + '/inserir', tarefa);
  }

  public alterar(tarefa: Tarefa): Observable<boolean> {
    return this.httpClient.put<boolean>(this.API + '/alterar', tarefa);
  }

  public excluir(id: number): Observable<boolean> {
    return this.httpClient.delete<boolean>(this.API + '/excluir/' + id);
  }

  public consultarPorId(id: number): Observable<Tarefa> {
    return this.httpClient.get<Tarefa>(this.API + '/' + id);
  }

  public consultarTodos(): Observable<Array<Tarefa>> {
    return this.httpClient.get<Array<Tarefa>>(this.API + '/listar');
  }

  public consultarPorFiltro(seletor: TarefaSeletor): Observable<Array<Tarefa>> {
    return this.httpClient.post<Array<Tarefa>>(this.API + '/filtro', seletor);
  }

  public criarTarefaAPartirDeTemplate(dto: TarefaTemplateDTO):Observable<Tarefa> {
    return this.httpClient.post<Tarefa>(this.API + '/inserir-template', dto);
  }

  public consultarItensDaTarefa(idTarefa: number): Observable<Array<ItemTarefa>> {
    return this.httpClient.get<Array<ItemTarefa>>(this.API + '/listar-itens/' + idTarefa);
  }

}
